src/placement/jump_map.c

```diff
@@ -409,8 +415,39 @@ get_target(struct pool_domain *curr_dom, struct pool_target **target,
 			end_dom = start_dom + (num_doms - 1);
 
 			range_set = isset_range(dom_used, start_dom, end_dom);
-			if (range_set)
-				clrbit_range(dom_used, start_dom, end_dom);
+			if (range_set) {
+				int idx;
+
+				/* Skip the domain whose targets are used up */
+				for (idx = start_dom; idx <= end_dom; ++idx) {
+					if (isclr(dom_occupied, idx))
+						clrbit(dom_used, idx);
+				}
+				/* if all children of the current dom have been
+				 * used, then let's go back its parent to check
+				 * its siblings.
+				 */
+				if (curr_dom != root_pos) {
+					setbit(dom_used, curr_dom - root_pos);
+					D_ASSERT(top != -1);
+					curr_dom = dom_stack[top--];
+				} else {
+					curr_dom = root_pos;
+				}
+				continue;
```

Comment:
Because this code is here, this should prevent iterating into this domain's children (targets), correct? I think this is the core/important logic. 

The logic immediately after this spot still seems to be needed to - `isset_range(dom_occupied)` - because of multi level. I just think the block with the first comment here might be unnecessary / could somehow result in an infinite loop. It seems like it should be an error or even assert?

Reply:
Resolved via Skype, just thinking about it.

Reply:
Won't this run into a dead loop if all domains are occupied? Please be aware this get_target() could be called from remapping code path, where all the bit arrays are already filled.

Reply:
hmm, you are right. good catch! thanks. dom_occupied check should be moved ahead.